tools/ci-build/release-dryrun/src/main.rs

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::time::Duration;
+
+use anyhow::{bail, Result};
+use clap::Parser;
+use indicatif::{ProgressBar, ProgressStyle};
+
+#[derive(Parser, Debug)]
+struct DryRunSdk {
+    /// Path to the aws-sdk-rust repo
+    #[clap(long)]
+    pub sdk_path: PathBuf,
+
+    /// Tag of the aws-sdk-rust repo to dry-run against
+    #[clap(long)]
+    pub rust_sdk_tag: String,
+
+    /// Path to the artifact produced by the release-dry run
+    #[clap(long)]
+    pub smithy_rs_release: PathBuf,
+}
+
+#[derive(Parser, Debug)]
+#[clap(
+    name = "release-dryrun",
+    about = "CLI tool to recursively update SDK/Smithy crate references in Cargo.toml files",
+    version
+)]
+#[allow(clippy::enum_variant_names)] // Want the "use" prefix in the CLI subcommand names for clarity
+enum Args {
+    DryRunSdk(DryRunSdk),
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+    match args {
+        Args::DryRunSdk(args) => dry_run_sdk(args)?,
+    }
+    Ok(())
+}
+
+fn step<T>(message: &'static str, step: impl FnOnce() -> Result<T>) -> Result<T> {
+    let spinner = ProgressBar::new_spinner()
+        .with_message(message)
+        .with_style(ProgressStyle::with_template("{spinner} {msg} {elapsed}").unwrap());
+    spinner.enable_steady_tick(Duration::from_millis(100));
+    let result = step();
+    let check = match &result {
+        Ok(_) => "✅",
+        Err(_) => "❌",
+    };
+    spinner.set_style(ProgressStyle::with_template(&format!("{{msg}} {{elapsed}}")).unwrap());
+    spinner.finish_with_message(format!("{check} {message}"));
+    result
+}
+
+fn dry_run_sdk(args: DryRunSdk) -> Result<()> {
+    step("Checking out SDK tag", || {
+        Command::new("git")
+            .arg("checkout")
+            .arg(&args.rust_sdk_tag)
+            .current_dir(&args.sdk_path)
+            .output()?;
+        Ok(())
+    })?;
+
+    step("Applying version-only dependencies", || {
+        Command::new("sdk-versioner")
+            .args([
+                "use-version-dependencies",
+                "--versions-toml",
+                "versions.toml",
+                "sdk",
+            ])
+            .current_dir(&args.sdk_path)
+            .output()?;
+
+        Command::new("git")
+            .args(["checkout", "-B", "smithy-release-dryrun"])
+            .current_dir(&args.sdk_path)
+            .output()?;
+        Command::new("git")
+            .args([
+                "commit",
+                "-am",
+                "removing path dependencies to allow patching",
+            ])
+            .current_dir(&args.sdk_path)
+            .output()?;
```

Comment:
These errors are going to be pretty useless when they happen.

Reply:
Maybe the message passed to `spinner.finish_with_message` can be augmented with the error message in case `step()` returns an error.

Reply:
good call, I fixed them all up so they're nice:
```
❌ Checking out SDK tag 0s                                                                                                                                                                                    
Error: failed to checkout aws-sdk-rust revision

Caused by:
    command `cd "/Users/rcoh/code/aws-sdk-rust" && "git" "checkout" "release-2023-12-1"` failed:
    error: pathspec 'release-2023-12-1' did not match any file(s) known to git
```